man/logmsg.3
.TH qlibs: logmsg 3 
.SH NAME
logmsg \- handle system errors and application log messages
.SH SYNTAX
.B #include \(dqlogmsg.h\(dq

int logmsg(const char *who, int ecode, unsigned int classs, char *msg)

\fBerr_sys\fR(w,e) logmsg(w,e,FATAL,"")
.br
\fBerr_sys_plus\fR(w,e,m) logmsg(w,e,FATAL,m)
.br
\fBerr_tmp\fR(w,e) logmsg(w,e,WARN,"")
.br
\fBerr_tmp_plus\fR(w,e,m) logmsg(w,e,WARN,m)
.br
\fBerr_int\fR(w,e,c) logmsg(w,e,c,"")
.br
\fBerr_int_plus\fR(w,e,c,m) logmsg(w,e,c,m)
.br
\fBlog_who\fR(w,m) logmsg(w,0,LOG,m)
.br
\fBlog_anon\fR(m) logmsg("",0,LOG,m)
.SH DESCRIPTION
\fBlogmsg\fR prints error, warning, or info/logging messages to stderr 
and potentially terminates the calling program, depending on the \fIclass\fR given.
\fIwho\fR is the name of the program, \fIecode\fR is an error code, 
\fIclass\fR determines the behavior upon call and \fImsg\fR is the logging message. 
Read "error.h" to learn related constants.
.SH ECODE
\fIecode\fR is the error code and subject to be displayed in the log file and
potentially used upon exit in case the \fIclass\fR equals \fBFATAL\fR or \fBDROP\fR.

To avoid conflicts with syscall error codes, appplication defined error codes should be negative. 
The values \fI-15\fR, \fI-100\fR and \fI-111\fR are reserved for backwards compatibility. 
.SH CLASS
The \fIclass\fr parameter indicates how the application handles exceptions and displays the
log message.
.TP 4
o
\fBLOG\fR, \fBINFO\fR, \fBALERT\fR, \fBWARN\fR - display message and continue operation
.TP 4
o
\fBDROP\fR - display warning message and continue while going to the calling program
.TP 4
o
\fBUSAGE\fR, \fBSYNTAX\fR, \fBFATAL\fR, \fBERROR\fR 
- display error message and exit application with error code
.RE

\fBINFO\fR, \fBALERT\fR, \fBWARN\fR, \fBDROP\fR, \fBUSAGE\fR, and \fBFATAL\fR as well 
as \fBERROR\fR display the respective class string like \fIwarning:\fR in the log message, 
while \fBLOG\fR shows the log message only.

The class \fBFATAL\fR should be used for system error codes only, rather \fBERROR\fR
and \fBWARN\fR shall be set in conjunction with an application error/warning.
.SH MESSAGE
If the custom message \fImsg\fR is given, it will be printed additionally.
.SH FORMAT
The log message format consists of the tokens
\fIwho\fR: (\fIecode\fR) \fImsg\fR : \fImsg\fR.
.I ecode 
is displayed only for classes \fBFATAL\fR, \fBERROR\fR, or \fBDROP\fR.
.I msg
is the system's explanation according to the variable
.I errno
if provided.
.SH NOTES
.I logmsg.c 
uses 
.I errstr.c 
routines.
Error codes and classes are defined in 
.I error.h 
and included by 
.IR logmsg.h .
.SH "EXIT CODE"
\fBlogmsg\fR exits \fIecode\fR for classes \fBERROR\fR, \fBFATAL\fR, \fBSYNTAX\fR, 
and \fBUSAGE\fR terminating the application.
.SH HISTORY
Dan Bernstein used sets of \fIstrerr_dieY*()\fR and \fIstrerr_warnY()\fR messages 
which explicitely determine the message and behavior class. 
Other classes were occasionally defined on demand, such als \fIusage()\fR. 

Kai Peter introduced the \fIerrmsg\fR facility in his \fBqlibs\fR 
including a \fBsyslog\fR compliant \fIseverity\fR as second parameter. 
.SH EXAMPLES
The macro definitions uses  \fBw\fR for the calling program, 
\fBe\fR for error code, \fBc\fR for class, and \fBm\fR for message.

  #include "logmsg.h"
  #define WHO "my_prog"

  err_sys(WHO,errno);
  err_sys_plus(WHO,-111,"additional message");

  err_tmp("",-100);
  err_tmp("",errno,"additional message");

  log_who(WHO,"message");

log_anon() is like log_who() but doesn't print the caller name.

An user defined message \fBs\fR can be build from multiple arguments by using the \fIB\fR
(build) macro:

  err_sys_plus((errno),B("unable to run: ",*argv));
.SH "SEE ALSO"
syslog(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
	DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
	DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.


logmsg.c
#include <unistd.h>
#include <stdarg.h>
#include <errno.h>
#include "buffer.h"
#include "fmt.h"
#include "str.h" 
#include "stralloc.h" 
#include "logmsg.h"

/**
	@file logmsg.c
	@author kp, feh
	@source qlibs
	@brief unified system and error message handling
*/

#define WHO "logmsg"

char *build_log_msg(const char *x[])
{
  stralloc sa = {0};
  stralloc_copys(&sa,"");  

  while(*x) { if (!stralloc_cats(&sa,*x++)) err_sys(WHO,errno); } /* concatenate *x */
  
  if (!stralloc_0(&sa)) err_sys(WHO,errno);
  return(sa.s);
}

void logmsg(const char *who,int ecode,unsigned int class,const char *msg)
{
  char strnum[FMT_ULONG];
  char *codestr = "";
  char *classstr = "";
  char *errmsg = "";

  errno = 0;   /* re-initialize errno, value is in 'code' now */

/* Part 1: obtain the (error) code -- perhaps received from OS */

  if (ecode != 0) {
    codestr = "";
    if (ecode < 0) {           // check for negative error codes 
      ecode = (ecode^-1) + 1;
      codestr = "-";
    }
    strnum[fmt_ulong(strnum,ecode)] = 0;    /* format for output */
    char *temp = strnum;
    codestr = str_cat(codestr,temp);
  }

/* Part 2: behavioral on error */

  switch (class) {
    case ERROR:  classstr = "error: ";    break;   // info + exit
    case FATAL:  classstr = "fatal: ";    break;   // info + exit 
    case DROP:   classstr = "drop: ";     break;   // info + next call/iteration
    case ALERT:  classstr = "alert: ";    break;   // info + next statement
    case WARN:   classstr = "warning: ";  break;   // info + next statement
    case INFO:   classstr = "info: ";     break;   // info + continue
    case SYNTAX: classstr = "syntax: ";   break;   // info + exit
    case USAGE:  classstr = "usage: ";    break;   // info + exit
    case TEMP:   classstr = "temp: ";    break;   // info + exit
    case CAT:    classstr = "";	          break;   // info w/o \n
    default: 
      class = LOG; classstr =  "";        break;   // custom info + continue
  }

/* Part 3: get system error message */

  if (class == FATAL || class == DROP)
    errmsg = error_str(errno);

/* Part 4: construct log message: Source: Class (Ecode) Message: Errmsg */

  buffer_puts(buffer_2,who);
  buffer_puts(buffer_2,": ");
  buffer_puts(buffer_2,classstr);
  if (class == FATAL || class == DROP || class == ERROR) {
    buffer_puts(buffer_2,"(");
    buffer_puts(buffer_2,codestr);
    buffer_puts(buffer_2,") ");
  }
  buffer_puts(buffer_2,msg);
  if (errno)  {
    buffer_puts(buffer_2,": ");
    buffer_puts(buffer_2,errmsg);
  }
  if (class != CAT) {
    buffer_puts(buffer_2,"\n");
    buffer_flush(buffer_2);
  }

  if (class == USAGE) _exit(USAGE); 
  if (class == SYNTAX) _exit(SYNTAX); 
  if (class == FATAL || class == ERROR) _exit(ecode);
}
